Introduce useOverrideFields plugin

santino opened the pull request:
## Description

This PR implements a new plugin to override fields’ resolvers.

The goal is to replace the default resolver with something designed as a drop-in function that prevents the original resolver logic from being executed and eventually returns a simple value instead.
It is mostly meant to be for empty values, such as empty strings, null or undefined; but it can be fully customised to be anything more complex, such as returning values from a cache.

Fields can be targeted by plain string or Regular Expressions.

Targeted fields are always overridden by default, but custom logic can be implemented with the built-in `shouldOverride` function, to enable/disable resolver function replacement.

As usual, feedback welcome. Once we’re happy with the implementation I will complete documentation and add tests.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

santino committed:
Initial implementation for OverrideFields plugin

santino committed:
Add changeset and additional case on README

dotansimha commented on packages/plugins/override-fields/README.md:
```diff
@@ -0,0 +1,50 @@
+const getEnveloped = envelop({
+  plugins: [
+    useOverrideFields({
+      fields: ['continents', 'country.callingCodes.countries', /country\.aplhaCode.*/],
```
I think this might need to be `Type.Field` instead of just field name? Otherwise you might effect types that you don't want to?

santino replied:
Yes, you're right.
It is something I did think about during development, but then I forgot to implement.

Will add that

santino committed:
Implement Type checking and complete documentation

santino committed:
Introduce tests for useOverrideFields plugin

santino committed:
Improve tests

adding mocks for original resolvers to make sure these are not hit when overriding
also reusing Type (Country) in multiple places within the schema to make sure targeting Type or nested fields works as expected

dotansimha committed:
adjust tests

n1ru4l reviewed (commented):
To me, it seems like this is might actually better be considered as schema tooling and I am not sure whether it should be an envelop plugin or standalone schema transform/mapper implemented via https://www.graphql-tools.com/docs/api/modules/utils#mapschema The performance of using mapSchema will probably even be better 🤔

santino committed:
Merge branch 'dotansimha:main' into override-fields
